mlflow/crewai/__init__.py

```diff
@@ -0,0 +1,75 @@
+"""
+The ``mlflow.crewai`` module provides an API for tracing CrewAI AI agents.
+"""
+
+import logging
+
+from packaging.version import Version
+
+from mlflow.crewai.autolog import (
+    patched_class_call,
+)
+from mlflow.utils.annotations import experimental
+from mlflow.utils.autologging_utils import autologging_integration, safe_patch
+
+_logger = logging.getLogger(__name__)
+
+FLAVOR_NAME = "crewai"
+
+
+@experimental
+@autologging_integration(FLAVOR_NAME)
+def autolog(
+    log_traces: bool = True,
+    disable: bool = False,
+    silent: bool = False,
+):
+    """
+    Enables (or disables) and configures autologging from CrewAI to MLflow.
+    Note that asynchnorous APIs and Tool calling are not recorded now.
+
+    Args:
+        log_traces: If ``True``, traces are logged for CrewAI agents.
+            If ``False``, no traces are collected during inference. Default to ``True``.
+        disable: If ``True``, disables the CrewAI autologging. Default to ``False``.
+        silent: If ``True``, suppress all event logs and warnings from MLflow during CrewAI
+            autologging. If ``False``, show all events and warnings.
+    """
+    # TODO: Handle asynchronous tasks and crew executions
+    # TODO: Tool calling is not supported since the interface of tool in CrewAI is
+    # changing drastically. Add patching once it's stabilized
+    import crewai
+
+    try:
+        class_method_map = {
+            crewai.Crew: ["kickoff", "kickoff_for_each", "train"],
+            crewai.Agent: ["execute_task"],
+            crewai.Task: ["execute_sync"],
+            crewai.LLM: ["call"],
+            crewai.Flow: ["kickoff"],
+            crewai.agents.agent_builder.base_agent_executor_mixin.CrewAgentExecutorMixin: [
+                "_create_long_term_memory"
+            ],
+        }
+        if Version(crewai.__version__) >= Version("0.83.0"):
+            # knowledge and memory are not available before 0.83.0
+            class_method_map.update(
+                {
+                    crewai.memory.ShortTermMemory: ["save", "search"],
+                    crewai.memory.LongTermMemory: ["save", "search"],
+                    crewai.memory.UserMemory: ["save", "search"],
+                    crewai.memory.EntityMemory: ["save", "search"],
+                    crewai.Knowledge: ["query"],
+                }
+            )
+
+        for cls, methods in class_method_map.items():
+            for method in methods:
+                safe_patch(
+                    FLAVOR_NAME,
+                    cls,
+                    method,
+                    patched_class_call,
+                )
+    except AttributeError as e:
+        _logger.error("An exception happens when applying auto-tracing to crewai. Exception: %s", e)
```

Comment:
Shall we try-catch each method patching independently? E.g., when a class `Knowledge` is renamed, should we stop tracing at all or still trace other classes?

Reply:
Yeah, I was in two minds. If we implement it in that way, we cannot define `class_method_map` since an AttributeException is raised immediately when the map is defined.

Reply:
What about making `class_method_map` key string and import the class one by one within each try-catch?

Reply:
Let me try that direction!